js/chatbot/shared/ChatBot.svelte

```diff
@@ -222,11 +223,24 @@
 		message: NormalisedMessage,
 		selected: string | null
 	): void {
+		const val_ = value as NormalisedMessage[];
 		if (selected === "undo") {
-			dispatch("undo");
+			let last_index = val_.length - 1;
+			while (val_[last_index].role === "assistant") {
+				last_index--;
+			}
+			dispatch("undo", {
+				index: val_[last_index].index
+			});
 			return;
 		} else if (selected === "retry") {
-			dispatch("retry");
+			let last_index = val_.length - 1;
+			while (val_[last_index].role === "assistant") {
+				last_index--;
+			}
+			dispatch("retry", {
+				index: val_[last_index].index
+			});
```

Comment:
And what's happening here? Why do we need a while loop, can't we just look at the index of the message that the retry/undo button is attached to?

Reply:
The message that got clicked is the bot message. But you want to retry/undo the user message. So we have to find the last user message that got submitted and send that index to the server. That's what the while loop is doing but there could be a more idiomatic functional programming approach to do that.